subsys/bluetooth/audio/bap_unicast_server.c

```diff
@@ -28,6 +29,17 @@ LOG_MODULE_REGISTER(bt_bap_unicast_server, CONFIG_BT_BAP_UNICAST_SERVER_LOG_LEVE
 
 static const struct bt_bap_unicast_server_cb *unicast_server_cb;
 
+int bt_bap_unicast_server_register(const struct bt_bap_unicast_server_register_param *param)
+{
+	LOG_DBG("reg");
+	return bt_ascs_service_register(param->snk_cnt, param->src_cnt);
+}
```

Comment:
> Alternatively we can merge the two so that bt_bap_unicast_server_register also takes a callback structure, as there can only ever be a single Unicast Server, right?

Isn't this kind of what I did before the latest changes? Except that everything was stored in the _cb function? :)

I think that we can keep them separated, on the basis that there might be some usecases where you want to un/reregister callbacks without shutting down the service(?). I dont really see a concrete use-case in front of me though.

Reply:
Indeed. The difference is that function was called "register_cb" which didn't fit what the function did. 

We can keep them separate for now, but if both always need to be called to ensure any functionality, then we may want to merge them.

Reply:
The issue with this is that 'bt_ascs_init' is dependent on the callback, since this is where ascs gets the handle to the unicast callbacks. It also reminds me that this hit when when doing this implementation in the first version. Also, ASCS cant be used without callbacks registered, as I mentioned here:
>That is fair enough, and I definetly see your point. The reason it was added here is that 'bt_bap_unicast_server_register_cb' initializes ascs, making ascs unusable unless you actually call cb registration first (wich makes sense, since some state machine transition needs input from user anyway).

So in a way, it does make sense that callback registration initializes ASCS.

However, can it make sense to keep registration and initialization separated? Technically, I guess there can be some use-case (which I can't concretely think about right now), where the user wants to register the service in the gatt db, even though callbacks are not registered yet. 
The other benefit would be that the user more easily can un/reregister callbacks without unregistering the service (prompting gatt db change notification, etc).

I vote for keeping it as is - but I'm open for changing it, i.e. merging them - but if so I would prefer to do that in an upcoming commit instead.

Reply:
Those are good points. Given the implementation of `bt_ascs_init` it's kind of weird. 

ASCS is today built like it depends on the unicast server (due to the dependency on the unicast server callback structure), but in reality it should be the other way around. The unicast server should register an ASCS callback structure, so that ASCS can call the callback (if registered), regardless of whether it's the BAP unicast server or some other implementation. This is, however, a larger refactor way outside the scope of this PR. 

Given how `bt_ascs_init` is currently defined, and how refactoring this entire part is non-trivial, then I'm leaning towards putting the CB registration into the `bt_bap_unicast_server_register` function, so that `bt_bap_unicast_server_register` can call `bt_ascs_init`. Registering callbacks (`bt_bap_unicast_server_register_cb`) should not be responsible for initializing a service. 

Given the complexity I'll leave it up to you whether to have one or 2 functions, but if we do have 2, then we need to add checks to ensure that they are called in the right order

Reply:
I will stick with keeping them separated for, mainly due to the complexity of the changes - and how severe the breakage of the ascs unit tests tends to become.
Ill see if I can get some time to work on merging them after this is completed, otherwise Ill create an issue instead.